llama/sampling_ext.cpp

```diff
@@ -45,25 +46,22 @@ llama_token common_sampler_csample(struct common_sampler *sampler, struct llama_
     return common_sampler_sample(sampler, ctx, idx);
 }
 
-int schema_to_grammar(const char *json_schema, char *grammar, size_t max_len)
+struct parsed_grammar * schema_to_grammar(const char *json_schema)
 {
+    struct parsed_grammar *result = new parsed_grammar();
     try
     {
         nlohmann::ordered_json schema = nlohmann::ordered_json::parse(json_schema);
         std::string grammar_str = json_schema_to_grammar(schema);
-        size_t len = grammar_str.length();
-        if (len >= max_len)
-        {
-            len = max_len - 1;
-        }
-        strncpy(grammar, grammar_str.c_str(), len);
-        return len;
+        result->length = grammar_str.length();
+        result->grammar = new char[result->length];
```

Comment:
We'd have to manage the memory in this as well which as grammar or the result struct would not be getting freed. Probably better to maintain the signature from before (we'd also have to update the sampling code otherwise)

Reply:
Both `result` and `result->grammar` are freed in `SchemaToGrammar`, lines 584 and (in case `length>0` which is when we have a grammar) line 593:
https://github.com/ollama/ollama/blob/21a4811c9c69e7a9d4c36af94729717129371ad9/llama/llama.go#L578-L595